### ISSUE TITLE
Concerns regarding hosts file and local web server

### ISSUE THREAD

**Issue by forthrin** [StevenBlack/hosts#1974]
Thanks for a great "nip it in the bud"-project! I wanted to share some relevant concerns:

1. How do I keep 0.0.0.0 from triggering a locally run web server?
2. Why do DNS servers return 127.0.0.1 for certain domains?
3. What happens if the localhost/loopback mappings do not come first in the hosts file?

Note: [0.0.0.0 is assigned specific meanings in a number of contexts](https://en.wikipedia.org/wiki/0.0.0.0)

```
$ curl --verbose http://0.0.0.0/
* Connected to 0.0.0.0 (127.0.0.1) port 80
< Server: nginx

$ dig istole.it
istole.it.		1808	IN	A	127.0.0.1

$ less /etc/hosts
93.184.216.34 example.com
127.0.0.1 localhost
255.255.255.255 broadcasthost
93.184.216.34 example.org
```

**Comment by StevenBlack** [comment]
Hi @forthrin thank you for your questions.

> 1. How do I keep 0.0.0.0 from triggering a locally run web server?

What OS are you running?  You probably should not be running a webserver on `0.0.0.0`

> 2. Why do DNS servers return 127.0.0.1 for certain domains?

Which DNS Servers?  Which domains?

> 3. What happens if the localhost/loopback mappings do not come first in the hosts file?

I think it's in the spec: those mappings must be listed first.  if not, then the behavior is undefined. Therefore nobody knows what happens because that will be implementation specific.

**Comment by ScriptTiger** [comment]
> $ curl --verbose http://0.0.0.0/
> * Connected to 0.0.0.0 (127.0.0.1) port 80
> < Server: nginx

As far as the cURL results, you can clearly see cURL is remapping 0.0.0.0 to 127.0.0.1, which is also clearly your localhost designation in your hosts file, as would be expected. As to why cURL is doing that, you would have to look further into the specific cURL implementation you're using, or perhaps 0.0.0.0 is being resolved to 127.0.0.1 by something else poorly configured in the network chain.

Does 0.0.0.0 get remapped by other applications on the same system, such as a Web browser? If 0.0.0.0 is getting remapped, because obviously 0.0.0.0 is not routable, than that raises a host of other questions as to what else is potentially getting hijacked or leaking on your system due to faulty remapping. There could easily be a virus syphoning traffic through a tunnel somewhere you didn't even know you had. Arbitrarily remapping reserved non-routable addresses is clearly not expected behavior and you have a non-standard issue which deviates from the norm and may be difficult to speculate on, but that you should definitely resolve as soon as possible due to it being an obvious security concern of extremely high risk potential, without exaggeration.

I can tell you from experience, over the last 20 years or so I've hosted many a Web server on Linux and Windows systems, no Macs yet (sorry, @StevenBlack), I commonly use the 0.0.0.0 as a black hole address in host files and have yet to have any issues. As @StevenBlack also said, as well as any best practice will tell you, using the 0.0.0.0 as a configuration in your Web server is a terrible idea. You should always bind statically to a specific address/interface and never just to all available addresses, it's a huge security concern. When you're dealing with anything that touches the Internet/a public network, always, always, always statically configure as much as you can and don't get lazy. Being lazy is what gets most people because they don't know what's going on in those lapses.

**Comment by ler762** [comment]
On 6/1/22, ScriptTiger  wrote:
>> $ curl --verbose http://0.0.0.0/
>> * Connected to 0.0.0.0 (127.0.0.1) port 80
>> < Server: nginx
>
> As far as the cURL results, you can clearly see cURL is remapping 0.0.0.0 to
> 127.0.0.1, which is also clearly your localhost designation in your hosts
> file, as would be expected. As to why cURL is doing that, you would have to
> look further into the specific cURL implementation you're using, or perhaps
> 0.0.0.0 is being resolved to 127.0.0.1 by something else ...

My understanding is that 0.0.0.0 as a destination is the host machine:
https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml

0.0.0.0/32 	"This host on this network" 	[RFC1122], Section 3.2.1.3

**Comment by ScriptTiger** [comment]
@ler762, please read the information you provided carefully. `0.0.0.0/32` is CIDR notation for a network block. As anyone that knows IPv4 can tell you, the first address of every network is automatically not routable because it's used as a network ID, as the last IP is also not usable because it's used as a broadcast address. This means no matter how you want to chop of the entirety of the IPv4 address space into networks, 0.0.0.0 will always and forever be not routable. IPv6 changes this and drops both network IDs and broadcast addresses, where broadcast addresses are redefined within the new multicast space.

**Comment by ler762** [comment]
On 6/1/22, ScriptTiger ***@***.***> wrote:
> @ler762, please read the information you provided carefully. `0.0.0.0/32` is
> CIDR notation for a network block.

/32 is CIDR notation for the first 32 bits of an address - i.e. an
IPv4 host address
0.0.0.0/8 is the entire 0.x.x.x address block (i.e. 1st 8 bits are 0)

> As anyone that knows IPv4 can tell you,
> the first address of every network is automatically not routable because
> it's used as a network ID, as the last IP is also not usable because it's
> used as a broadcast address.

I guess you've never used a /31 on point-to-point links :)

> This means no matter how you want to chop of
> the entirety of the IPv4 address space into networks, 0.0.0.0 will always
> and forever be not routable.

Where did I say it was routable?

Lee

**Comment by forthrin** [comment]
> Hi @forthrin thank you for your questions.
> 
> > 1. How do I keep 0.0.0.0 from triggering a locally run web server?
> 
> What OS are you running? You probably should not be running a webserver on `0.0.0.0`

macOS. 0.0.0.0 maps to localhost. Unless one can change how the OS handles this or block the requests on any given web server (nginx in my case), the only solution I can see is changing `0.0.0.0` in the hosts file to eg. `0.0.0.1`. All IP addresses in the `0.*.*.*` range are reserved for the same purpose, and can be blocked by a web server with the exception of `0.0.0.0` because of its catch-all meaning. Is there any reason not to change this except aesthetics? 

> > 2. Why do DNS servers return 127.0.0.1 for certain domains?
> 
> Which DNS Servers? Which domains?

Any DNS server (eg. 1.1.1.1) and the domain `istole.it` as exemplified.

> > 3. What happens if the localhost/loopback mappings do not come first in the hosts file?
> 
> I think it's in the spec: those mappings must be listed first. if not, then the behavior is undefined. Therefore nobody can say what happens because that will be implementation specific.

I've defied this for a while, seemingly without tangible damage. Reason I wanted to concatenate the hosts file with custom records using a simple one-liner. Is there simple one-liner that inserts custom records in the correct place?

`sudo bash -c 'cat ~/.hosts <(curl -Sfs https://raw.githubusercontent.com/StevenBlack/<snip>/hosts) > /etc/hosts'`

**Comment by DavidCWGA** [comment]
> macOS. 0.0.0.0 maps to localhost. 

Mine doesn't.

```
$ ping 0.0.0.0
PING 0.0.0.0 (0.0.0.0): 56 data bytes
ping: sendto: Socket is not connected
```

No system should ever do:
https://en.wikipedia.org/wiki/0.0.0.0

**Comment by forthrin** [comment]
True:

```
$ ping 0.0.0.0
PING 0.0.0.0 (0.0.0.0): 56 data bytes
ping: sendto: Socket is not connected
```

However:

```
$ curl --head http://0.0.0.0/
HTTP/1.1 200 OK
Server: nginx
<snip>
```

And:

```
$ curl --head http://0.0.0.1/
curl: (7) Couldn't connect to server
```

**Comment by StevenBlack** [comment]
On my Mac laptop:

```
$ curl --head http://0.0.0.0/                                                                           23:18:28
curl: (7) Failed to connect to 0.0.0.0 port 80 after 7 ms: Connection refused
```

@forthrin it seems you have a system-local config issue that's out-of-scope for us here.

Closing.

**Comment by forthrin** [comment]
@StevenBlack: Do you have a local web server running on port 80? So that the example below does work on your machine? If so, you must be right about 0.0.0.0 being a local issue here. It does happen on several of my Macs, though. Wonder why.

```
$ curl --head http://127.0.0.1/
HTTP/1.1 200 OK
Server: nginx (or some)
```

PS! Do comment on public DNS servers returning 127.0.0.1 for some domains if you have any knowledge about this.

**Comment by StevenBlack** [comment]
@forthrin there's no mystery here.  Domains like `istole.it` have DNS records, in this case an `A` record, that points to `127.0.0.1`

```
$ dig istole.it                                                                                         
istole.it.		3599	IN	A	127.0.0.1
```

**Comment by forthrin** [comment]
@StevenBlack: Yes, but why? Is there any purposeful, non-malicious meaning to public DNS servers pointing domains to localhost IP addresses?

And please confirm that you do indeed have a local web server running with regard to the example you gave in your previous reply.

**Comment by ScriptTiger** [comment]
> @StevenBlack: Yes, but why? Is there any purposeful, non-malicious meaning to public DNS servers pointing domains to localhost IP addresses?

`istole.it` is a torrent tracker. It's up for debate whether you consider torrents malicious or not. Obviously, malicious torrents are malicious. However, it's not up for debate whether `istole.it` tracks torrents with illegal content, such as pirated material. So, they probably obfuscate their records for obvious reasons of evasion, which is completely irrelevant to how torrent swarms, which are peer-to-peer, utilize their services.

> How do I keep 0.0.0.0 from triggering a locally run web server?

Properly configure your Web server to bind/listen to specific addresses/interfaces and do not specify `0.0.0.0` as one such address. If you are just binding/listening to all addresses, you can consider your Web server misconfigured. This repository does not offer support for one-off misconfigured services, nor should it.

**Comment by ler762** [comment]
On 6/11/22, forthrin wrote:
> @StevenBlack: Yes, but why? Is there any purposeful, non-malicious meaning
> to public DNS servers pointing domains to localhost IP addresses?

net-snmp uses it for testing::

# 408 Lookup tests, require network access
# 408a Test lookup returning a single host
CONFIGAGENT 'com2sec t408a onea.net-snmp.org c408a'
# 408a Test lookup returning multiple hosts
CONFIGAGENT 'com2sec t408b twoa.net-snmp.org c408b'

$ dig +short onea.net-snmp.org
127.0.0.1

$ dig +short twoa.net-snmp.org
127.0.0.3
127.0.0.2

**Comment by forthrin** [comment]
> > How do I keep 0.0.0.0 from triggering a locally run web server?
> 
> Properly configure your Web server to bind/listen to specific addresses/interfaces and do not specify `0.0.0.0` as one such address. If you are just binding/listening to all addresses, you can consider your Web server misconfigured. This repository does not offer support for one-off misconfigured services, nor should it.

All `listen` entries in `nginx.conf` are bound to 127.0.0.1, confirmed by `lsof -i`. Still `http://0.0.0.0/` gets through to the web server via `nginx`. I realize this is bordering on out of scope, unless it is a common mistake or default setting people should be aware of when using a hosts file.

**Comment by ScriptTiger** [comment]
Have you tried creating a virtual adapter and assigning it a private IP address that is neither a loopback nor overlapping with your live network, like something in the 192.168.0.0/16 or 10.0.0.0/8 blocks, and then assigning your Web server to listen on that address? It should offer you the same level of security, as long as you don't route that private IP address out, as using a loopback, but without the headache Mac seems to be giving you. Again, as long as it's not overlapping with your live network. So, if your live network is 192.168.0.1/24, then obviously don't use an IP in that network that could be potentially routable.

**Comment by forthrin** [comment]
Thanks for your kind advice. This is somewhat outside my network knowledge, but I suppose I could find a guide and try. However, I tried a third mac, a pristine installation with no web server, and it does treat 0.0.0.0 and 0.0.0.1 differently. 

```
$ curl http://0.0.0.0/
curl: (7) Failed to connect to 0.0.0.0 port 80 after 16 ms: Connection refused
$ curl http://0.0.0.1/
curl: (7) Couldn't connect to server
$ curl http://0.0.0.2/
curl: (7) Couldn't connect to server
```

I tried the same on an Alpine Linux VM, with similar results. So there is something about 0.0.0.0 that seems fundamentally different from all other addresses in the 0.* range. Unless someone would kindly step forward with true insight on the phenomenon, the pragmatic thing seems to just to do s/0.0.0.0/0.0.0.1/ on the hosts file and carry on the day.

**Comment by ScriptTiger** [comment]
Glad you figured this out! Any further follow-ups would probably be better suited for Stack Exchange. I think Steven probably already stopped following this thread a few comments ago.

I think you are fundamentally going about this entirely the wrong way, however. None of the tests you are performing have anything to do with the hosts file. When you `curl http://0.0.0.0/`, it's using your systems routing table, not the hosts file. Before connecting to a host, your OS parses the URL and knows the difference between an IP address and a domain name. Your OS only uses the hosts file to resolve domain names, not IP addresses, which is the job or your system's routing table.

Try this and post the results:
`curl https://analytics.google.com/`

**Comment by forthrin** [comment]
This I know very well. My point is that if there is uncertainty to how 0.0.0.0 is handled, and may under some circumstances be treated as localhost for some reason, and 0.0.0.1 is never or very unlikely treated as localhost, then 0.0.0.1 would be better used in the hosts file, so requests to blocked domains are sure to go nowhere than maybe to your local web server.

And yes, obviously your suggested request goes to my local web server, as long as that domain is mapped to 0.0.0.0 in the hosts file. Changing it to 0.0.0.1 makes it go nowhere like expected. I also had a friend do these tests this on his Mac and a Linux VM, with identical results as me. So it's not just me.

$ curl --head https://analytics.google.com/
Server: nginx/1.19.10

Trying to get this message across is an uphill battle with no clarity or resolution in sight, costing involved parties many more calories than doing my one-liner workaround. Which is also reminds me why I stopped using Stack Abuse years ago.

**Comment by ScriptTiger** [comment]
> $ curl --head https://analytics.google.com/
> Server: nginx/1.19.10

Can you confirm you have the black list correctly installed in your hosts file? If so, you obviously have a misconfigured Web server that's responding to literally everything. If you don't understand what your problem is by this point, I'm not sure how else to help you. Do you know what denial of service is? Or literally the thousands of other attacks you're opened to right now? Probably not a super great idea to be responding to literally everything. Or do you literally have a virtual host configured for `analytics.google.com`? Switch to Apache and try that.

> Trying to get this message across is an uphill battle with no clarity or resolution in sight, costing involved parties many more calories than doing my one-liner workaround. Which is also reminds me why I stopped using Stack Abuse years ago.

I don't see how doing anything with the hosts file is going to stop your Web server from responding to requests that are not even addressed to it. After you've done this, post your results for `curl http://0.0.0.0/` again. If all you've done is change your hosts file, your Web server will still be responding to completely arbitrary requests with completely arbitrary host headers.

**Comment by ler762** [comment]
First off, I think it's pretty cool that someone is validating that
their host file works the way they think it's supposed to work.

2nd, has this come up before?  If no, doesn't that seem a bit sad?

>> $ curl --head https://analytics.google.com/
>> Server: nginx/1.19.10

let's take a look at my saved file:

$ grep analytics.google unified-hosts.txt
0.0.0.0 analytics.google.com

So anyone incorporating the unified hosts file in their /etc/hosts and
doing a "curl --head https://analytics.google.com/" is really doing a
"curl --head https://0.0.0.0/" -- correct?

On my machine
$ curl 0
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>400 Bad Request</title>
</head><body>
  <.. snip ..>

and .. yes, apache is listening on port 80, all interfaces:
$ ss -nlt | grep ':80 '
LISTEN 0      511                *:80              *:*

So let's tell apache to _not_ listen to 0.0.0.0
/etc/apache2/apache2.conf has this snippet

# Include list of ports to listen on
Include ports.conf

and ports.conf has
Listen 80

There's probably a more correct way to do this, but for a POC, change
the line to
Listen <myIPaddress>:80
and restart apache

and .. yay! apache is now listening only on <my ip address>:80

$ curl 0
curl: (7) Failed to connect to 0 port 80: Connection refused

$ curl --head <my ip address>
HTTP/1.1 200 OK
  [.. snip ..]

So the problem is that apache is listening to port 80 on all interface
addresses; fix it to listen to port 80 only on the network card
interface address and problem solved.

Or use the OP's suggestion and change all the 0.0.0.0 addresses to
0.0.0.1 -- which I don't know what all the implications are, but it
would fix other things like

$ grep namezero /etc/hosts
0.0.0.0 namezero

$ ping namezero
PING namezero (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=0.032 ms
64 bytes from localhost (127.0.0.1): icmp_seq=2 ttl=64 time=0.047 ms
64 bytes from localhost (127.0.0.1): icmp_seq=3 ttl=64 time=0.047 ms
^C
--- namezero ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2052ms

  ---- edit /etc/hosts

$ grep namezero /etc/hosts
0.0.0.1 namezero

$ ping namezero
PING namezero (0.0.0.1) 56(84) bytes of data.
^C
--- namezero ping statistics ---
6 packets transmitted, 0 received, 100% packet loss, time 5124ms


Although I prefer a different solution: bind rpz; others prefer
pi-hole for blocking name -> address lookups

$ ping analytics.google.com.
ping: analytics.google.com.: Name or service not known

Lee

**Comment by StevenBlack** [comment]
I have three observations.

First observation: on my Mac:

```
curl --head https://analytics.google.com/                                                                                                
curl: (7) Failed to connect to analytics.google.com port 443 after 13 ms: Connection refused
```

**Therefore**, works as expected


```
curl --head https://0.0.0.0/                                                                                                
curl: (7) Failed to connect to analytics.google.com port 443 after 13 ms: Connection refused
```

**Therefore**, also works as expected.

Second observation: OP has misconfigured machine.

Third observation: hundreds of developers follow this repository and I want to apologize to all subscribers for this never-ending saga of OP's misconfigured Mac.  It ends now.  Locking this conversation.
